pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java

```diff
@@ -127,6 +145,19 @@ public int filterEntriesForConsumer(Optional<EntryWrapper[]> entryWrapper, int e
             msgMetadata = msgMetadata == null
                     ? Commands.peekMessageMetadata(metadataAndPayload, subscription.toString(), -1)
                     : msgMetadata;
+            if (CollectionUtils.isNotEmpty(entryFilters)) {
+                fillContext(filterContext, batchSizes, sendMessageInfo, indexesAcks, cursor, isReplayRead,
+                        msgMetadata, subscription);
+                EntryFilter.FilterResult result = getFilterResult(filterContext, entry);
+                if (EntryFilter.FilterResult.REJECT == result) {
+                    PositionImpl pos = (PositionImpl) entry.getPosition();
+                    entries.set(i, null);
+                    entry.release();
+                    subscription.acknowledgeMessage(Collections.singletonList(pos), AckType.Individual,
```

Comment:
If we have 100 entries to dispatch and 50 will be filtered out, we can improve here is only call the subscription.acknowledgeMessage() once.